ush/remap.sh

```diff
@@ -0,0 +1,329 @@
+#   nangles = 1:
+#      Rotate the specificed vectors accordingly; here it is assumed
+#      that a single angle (e.g., `theta`) defines the grid rotation
+#      angle; the units of `theta` are assumed to be radians.
+#
+#   nangles = 2:
+#     Rotate the specified vectors accordingly; here it is assumed
+#     that the `cos(theta)` and `sin(theta)` have been computed
+#     a'priori and assigned the input file variable names defined by
+#     `cosang` and `sinang` respectively.
+#
+function cdo_rotate(){    
+    local varnames="${1}"
+    local varfile="${2}"
+    local interp_type="${3}"
+    local angles="${4}"  
+    local var_rotate_path="${PWD}/rotate.nc"
+    local nc_output_path="${PWD}/rotate_vars.nc"
+    
+    comma_split_string "${varnames}"    
+    varname_array=("${global_array[@]}")    
+    comma_split_string "${angles}"
+    angle_array=("${global_array[@]}")
+    strip_whitespace "${varname_array[0]}"
+    xvar="${out_string}"
+    strip_whitespace "${varname_array[1]}"
+    yvar="${out_string}"
+    nangles="${#angle_array[@]}"
+
+    # Interpolate the ocean vectors to the destination grid.
+    echo "Interpolating variable ${xvar} and ${yvar} to destination grid."
+    cdo_remap "${xvar}" "${varfile}" "${interp_type}" "${nc_output_path}" 
+    cdo_remap "${yvar}" "${varfile}" "${interp_type}" "${nc_output_path}"    
+    echo "Rotating and remapping variables ${xvar} and ${yvar} from file ${varfile}."
+
+    # This is for instances of rotated grid projections defined by a
+    # single array/attribute value (e.g., `theta`).
+    if (( nangles == 1 )); then
+        strip_whitespace "${angle_array[0]}"
+        theta="${out_string}"
+	cdo_remap "${theta}" "${varfile}" "${interp_type}" "${nc_output_path}"
+        cdo -expr,"xr=cos(${theta})*${xvar}+sin(${theta})*${yvar}; yr=cos(${theta})*${yvar}-sin(${theta})*${xvar}" -selname,"${xvar}","${yvar}","${theta}" "${output_path}" "${var_rotate_path}" >> "${REMAP_LOG}" 2>&1
+
+    # This is typically used for MOM6 (ocean model) forecast output;
+    # the angle of rotation `theta` is not explicitly provided by
+    # rather the cosine and sine of the respective angle.
+    elif (( nangles == 2 )); then
+        strip_whitespace "${angle_array[0]}"
+        cosang="${out_string}"
+	cdo_remap "${cosang}" "${varfile}" "${interp_type}" "${nc_output_path}" 
+        strip_whitespace "${angle_array[1]}"
+        sinang="${out_string}"
+	cdo_remap "${sinang}" "${varfile}" "${interp_type}" "${nc_output_path}" 	
+	cdo -expr,"xr=${cosang}*${xvar}+${sinang}*${yvar}; yr=${cosang}*${yvar}-${sinang}*${xvar}" -selname,"${xvar}","${yvar}","${cosang}","${sinang}" "${nc_output_path}" "${var_rotate_path}" >> "${REMAP_LOG}" 2>&1 
+	varname_update "xr" "${xvar}" "${var_rotate_path}"
+	cdo_remap "${xvar}" "${var_rotate_path}" "${interp_type}" "${output_path}"
+	varname_update "yr" "${yvar}" "${var_rotate_path}"
+	cdo_remap "${yvar}" "${var_rotate_path}" "${interp_type}" "${output_path}"
```

Comment:
For the vector remapping, it is suggested to do the following so that there is more consistency with the NCL/Python remapping programs:

1. Replace missing values with 0.0 before any interpolation or vector rotation. 
2. On the interpolated u and v fields, for grid points where the interpolated mask equals 1 (non-ocean points), set those values on the interpolated u and v fields to a netcdf FillValue. This mask can be derived from a 2D and 3D scalar variable (e.g. SST for the 2D variable and Temperature for the 3D variable).  

I think once the above 2 items are added, the CDO remapping will be in great shape.

Reply:
There is no CDO feature that I am currently aware of that can handle masks in such a way. It may require some more time and attention to add these features. However, since these items are only suggestions and not required for the CDO-based remapping to fit in the global-workflow, these items can be addressed in a separate issue and future PR only if needed, and if there are no objections from anyone. As `remap.sh` currently stands, I still think the results are considerably close to the original NCL remapping program `ush/ocnpost.ncl`.

Reply:
I do not following this repo, so I apologize if I'm out-of sync. A basic question: How (or where) does CDO obtain the land mask for the destination grid? 

In the NCL/Python versions, the destination land mask can be thought of as an "output", meaning the destination land mask is given by the points where a source grid ocean point cannot be mapped fully to a destination grid point (ie, the destination point would be represented by a combination of a land and ocean point on the source grid). This is more and more important as the bathymetry closes in since the ocean model bathymetry will not in general match something like etopo5. 

This is an entirely separate issue than the polar seam. That is the result of the fact that the last latitude on the T-cell tripole grid is at 89.64 and so no T-cell grid point can be mapped to 90N.

Reply:
@DeniseWorthen It is unclear what CDO uses as a land mask for the destination grid by default. It does look like there is a way to create a mask file and apply it to the interpolated field such as in this example on the CDO forum: https://code.mpimet.mpg.de/boards/53/topics/10933

Here are the basics of what I gathered from this example so far:

1. Interpolate field: `cdo -remapbil,r360x180 infile.nc infile_r360x180.nc`
2.  create mask file with the same resolution as the destination grid: `cdo -f nc -remapbil,r360x180 -topo topo_r360x180.nc`
3. mask the land areas using the mask file: `cdo -expr,'topo = ((topo<0.0)) ? 1.0 : topo/0.0' topo_r360x180.nc mask_land.nc`
4. mask interpolated field with the mask file: `cdo -mul mask_land.nc infile_r360x180.nc infile_r360x180_mask_land.nc`

Reply:
Thanks @EricSinsky-NOAA I took some time and rewrote the entire original ocnpost.ncl routine in Fortran. It reproduces the fields created by the original NCL to O~-8 in all fields: https://github.com/DeniseWorthen/ocnpost90

I wasn't asked to do this, but a fortran code is a) portable b) fast and c) debuggable.

Reply:
@DeniseWorthen Thanks for that. I suspect this will be what we ultimately go with. Once Rahul is back on Weds we will discuss the best way to proceed.

Thanks to everyone who's been working on this, particularly those who wrote stuff that may not be used, which I know can be frustrating.

Reply:
@DeniseWorthen Thank you for writing your NCL code in FORTRAN! I do agree that FORTRAN is a great option too for this sort of application, particularly because of its efficiency and flexibility. At first glance, your FORTRAN ocnpost program looks very close to your original NCL ocnpost program.